### PR TITLE
Fix raw key mode in load-base-split

### LINKED ISSUE #9613: load based split can't handle raw kv well
### What version of TiKV are you using?
v5.0.0-rc

### Steps to reproduce
AutoSplitController always encode key as if it's transaction kv, which is wrong in raw kv.

https://github.com/tikv/tikv/blob/80582cb06f845a400090febb5cbaaf05250a2695/components/raftstore/src/store/worker/split_controller.rs#L378-L382

This can result in large region that can't be split. I suggest to always treat split key as internal keys just like all other observers/workers do.

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: close #9613  

Problem Summary:load based split can't handle raw kv well 

### What is changed and how it works?

What's Changed:
use the split key directly in split info and use the encoded key when collecting data in kv_get, kv_scan, coprocessor, and so on

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Test results show that load-base-split works well on requests such as kv_get, coprocessor, raw_kv, raw_scan.

#### coprocessor
`sysbench --table_size=100000 --range_size=100 --point_selects=0 --sum_ranges=0 --order_ranges=0 --distinct_ranges=0 --config-file=config oltp_read_only run`
image
image

#### kv_get and coprocessor
`sysbench --table_size=100000 --config-file=config oltp_read_only run`
image
image

#### kv_get

`sysbench --table_size=100000 --config-file=config oltp_point_select run`

image

image

#### raw_get

`./bin/go-ycsb run tikv -P workloads/workloada -p threadcount=200 -p tikv.type=raw  -p tikv.pd=xxx:2379 `

`"readproportion"="1"`

image
image

#### raw_scan

`./bin/go-ycsb run tikv -P workloads/workloada -p threadcount=200 -p tikv.type=raw  -p tikv.pd=xxx:2379 `

```
recordcount=10000000
operationcount=100000000
workload=core

readallfields=true

readproportion=0
updateproportion=0
scanproportion=1
insertproportion=0

requestdistribution=uniform
scanlengthdistribution=uniform
````
ps. split-contained-score=1.0

image

### Release note 
- Fix the issue that load based split can't handle raw kv well

### CODE REVIEW

**Comment by lhy1024** [comment]
@nolouch @BusyJay PTAL

**Comment by lhy1024** [comment]
@nolouch @BusyJay PTAL

**Comment by BusyJay** [comment]
Any test cases about the mixed raw/txn scenario?

**Comment by lhy1024** [comment]
> Any test cases about the mixed raw/txn scenario?

I tried running mixed txn/raw through go-ycsb and sysbench, but I found it was difficult to avoid both key conflicts.

**Comment by BusyJay** [comment]
How about implementing them as unit test case?

**Comment by lhy1024** [comment]
> How about implementing them as unit test case?

I added it. PTAL @BusyJay

**Comment by lhy1024** [comment]
@BusyJay @nolouch PTAL

File: components/raftstore/src/store/worker/split_controller.rs [hunk 1]
```diff
@@ -377,11 +375,10 @@ impl AutoSplitController {
 
             recorder.record(key_ranges);
             if recorder.is_ready() {
-                let key = recorder.collect(&self.cfg);
-                if !key.is_empty() {
+                if let Some(key) = recorder.collect(&self.cfg) {
                     let split_info = SplitInfo {
                         region_id,
-                        split_key: Key::from_raw(&key).into_encoded(),
+                        split_key: key.clone(),
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/split_controller.rs:381]: Why clone?

File: src/storage/mod.rs [hunk 2]
```diff
@@ -710,18 +707,18 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
...
         let res = self.read_pool.spawn_handle(
             async move {
-                if let Ok(start_key) = start_key
-                    .as_ref()
-                    .map(|k| k.to_raw())
-                    .unwrap_or_else(|| Ok(vec![]))
-                {
-                    let mut key = vec![];
+                if let Some(start_key) = &start_key {
+                    let mut key = &vec![];
```
Comment (BusyJay) [inline src/storage/mod.rs:711]: How about
```
let end_key = match &end_key {
    Some(k) => k.as_encoded().as_slice(),
    None => &[],
};
```

File: src/storage/mod.rs [hunk 3]
```diff
@@ -351,10 +355,9 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
             self.read_pool.spawn_handle(
                 async move {
                     for get in &requests {
-                        let key = get.key.to_owned();
                         let region_id = get.get_context().get_region_id();
                         let peer = get.get_context().get_peer();
-                        tls_collect_qps(region_id, peer, &key, &key, false);
+                        tls_collect_qps(region_id, peer, &get.key, &get.key, false);
```
Comment (BusyJay) [inline src/storage/mod.rs:360]: Why not put it at L373?

File: components/raftstore/src/store/worker/split_controller.rs [hunk 4]
```diff
@@ -231,7 +229,7 @@ impl Recorder {
         split_balance_score: f64,
         split_contained_score: f64,
         sample_threshold: i32,
-    ) -> Vec<u8> {
+    ) -> Option<Vec<u8>> {
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/split_controller.rs:232]: Why use `Option`?

Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:232]: to distinguish between vec![] and empty key

File: components/raftstore/src/store/worker/split_controller.rs (hunk 1 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:381]: done

File: src/storage/mod.rs (hunk 3 above)
Comment (lhy1024) [reply src/storage/mod.rs:360]: done

File: src/storage/mod.rs (hunk 2 above)
Comment (lhy1024) [reply src/storage/mod.rs:711]: done

File: components/raftstore/src/store/worker/split_controller.rs (hunk 4 above)
Comment (BusyJay) [reply components/raftstore/src/store/worker/split_controller.rs:232]: Why need to distinguish both? Split key can't be `vec![]`.

File: src/storage/mod.rs [hunk 5]
```diff
@@ -579,21 +578,17 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
...
-                    tls_collect_qps(
-                        ctx.get_region_id(),
-                        ctx.get_peer(),
-                        &start_key,
-                        &key,
-                        reverse_scan,
-                    );
+                let mut key = &vec![];
+                if let Some(end_key) = &end_key {
+                    key = end_key.as_encoded();
```
Comment (BusyJay) [inline src/storage/mod.rs:583]: Should follow the similar style with L708.

File: components/raftstore/src/store/worker/split_controller.rs [hunk 6]
```diff
@@ -494,29 +493,75 @@ mod tests {
         sc.sample_key(b"", b"d", Position::Contained);
     }
 
+    fn gen_read_stats(region_id: u64, key_ranges: Vec<KeyRange>) -> ReadStats {
+        let mut qps_stats = ReadStats::default();
+        for key_range in &key_ranges {
+            qps_stats.add_qps(region_id, &Peer::default(), key_range.clone());
+        }
+        qps_stats
+    }
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/split_controller.rs:502]: Missing line.

File: src/storage/mod.rs (hunk 5 above)
Comment (lhy1024) [reply src/storage/mod.rs]: done

File: components/raftstore/src/store/worker/split_controller.rs (hunk 6 above)
Comment (lhy1024) [reply components/raftstore/src/store/worker/split_controller.rs:502]: done

**Review by BusyJay — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**